File: src/main/java/pers/neige/neigeitems/action/node/impl/PapiParser.java
package pers.neige.neigeitems.action.node.impl;

import lombok.NonNull;
import lombok.val;
import org.bukkit.OfflinePlayer;
import org.jetbrains.annotations.Nullable;
import pers.neige.neigeitems.action.ActionContext;
import pers.neige.neigeitems.action.node.NodeParser;
import pers.neige.neigeitems.manager.BaseActionManager;

import static pers.neige.neigeitems.manager.HookerManager.requestPapi;

public class PapiParser extends NodeParser {
    public PapiParser(@NonNull BaseActionManager manager) {
        super(manager);
    }

    @Override
    public @NonNull String getId() {
        return "papi";
    }

    @Override
    public @Nullable String parse(
        @NonNull ActionContext context,
        @NonNull String params
    ) {
        val player = context.getCaster();
        if (!(player instanceof OfflinePlayer)) return null;
        val args = params.split("_", 2);
        return requestPapi((OfflinePlayer) player, args[0], args[1]);
    }
}

File: project/common/src/main/kotlin/pers/neige/neigeitems/section/impl/PapiParser.kt
package pers.neige.neigeitems.section.impl

import org.bukkit.OfflinePlayer
import org.bukkit.configuration.ConfigurationSection
import pers.neige.neigeitems.manager.HookerManager.requestPapi
import pers.neige.neigeitems.section.SectionParser
import pers.neige.neigeitems.utils.StringUtils.joinToString

/**
 * papi节点解析器(仅包含即时声明节点)
 */
object PapiParser : SectionParser() {
    override val id: String = "papi"

    override fun onRequest(
        args: List<String>,
        cache: MutableMap<String, String>?,
        player: OfflinePlayer?,
        sections: ConfigurationSection?
    ): String {
        // 相较于papi(player, "%${args.joinToString("_")}%"), 这种方式性能略有提升, 因为少遍历了一次字符串
        return player?.let {
            requestPapi(player, args[0], args.joinToString("_", 1))
        } ?: "<$id::${args.joinToString("_")}>"
    }
}